### PR TITLE
#21: nested returns inspection

### PR DESCRIPTION
**Pull request by ccostin93**

### CODE REVIEW

File: lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    final Node callee;
+
+    public NodeCaller(@Nonnull final Node callee) {
+        this.callee = callee;
+    }
+
+
+    @Override
+    public Node caller() {
+        return callee.getParentNode()
```
Comment (ccostin93) [inline lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java:19]: initially, I created this as a private static method, but then I wasn't sure on your opinion on private static methods. I saw you opened https://github.com/iwillfailyou/java-inspections/issues/23, but still not sure you want to allow private static in your codebase.

File: lib/src/main/java/com/iwillfailyou/javaparser/Caller.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.iwillfailyou.javaparser;
+
+import com.github.javaparser.ast.Node;
+
+public interface Caller {
```
Comment (ccostin93) [inline lib/src/main/java/com/iwillfailyou/javaparser/Caller.java:5]: Didn't have a better naming idea

File: lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java (hunk 1 above)
Comment (nikialeksey) [reply lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java:19]: Yeah, any private method is a reason for a new class

File: lib/src/main/java/com/iwillfailyou/javaparser/Caller.java (hunk 2 above)
Comment (nikialeksey) [reply lib/src/main/java/com/iwillfailyou/javaparser/Caller.java:5]: The main idea of this interface - provide a node, which would be found by some algorithm (defined in concrete implementation). May be `interface Search<T extends Node> { List<T> find(); }` and implementation `class ParentMethodOrLambda implements Search<NodeWithParameters<?>>`

Comment (ccostin93) [reply lib/src/main/java/com/iwillfailyou/javaparser/Caller.java]: fixed. went for `Optional<T> find();` instead if `List`.

File: lib/src/main/java/com/iwillfailyou/inspections/nomultiplereturn/multiplereturns/MultipleReturnViolations.java [hunk 3]
```diff
@@ -27,10 +25,21 @@
...
+                method.findAll(ReturnStmt.class)
+                    .stream()
+                    .collect(Collectors.groupingBy((final ReturnStmt returnStmt) -> new NodeCaller(returnStmt).caller()));
+
+            returnStmtMap.forEach((final Node parent, final List<ReturnStmt> returnStmts) -> {
+                if (returnStmts.size() > 1) {
+                    multipleReturns.add(new JavaMultipleReturn(parent, root));
+                } else {
+                    final ReturnStmt returnStmt = returnStmts.get(0);
+                    if (returnStmt.getChildNodes().size() > 0 && returnStmt.getChildNodes().get(0) instanceof ConditionalExpr) {
```
[the remark this answers is not in the document]
Comment (ccostin93) [reply lib/src/main/java/com/iwillfailyou/inspections/nomultiplereturn/multiplereturns/MultipleReturnViolations.java]: fixed

File: lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+                   .map((final Node parent) ->
+                            (parent instanceof LambdaExpr || parent instanceof MethodDeclaration)
+                                ? parent
+                                : new NodeCaller(parent).caller())
```
[the remark this answers is not in the document]
Comment (ccostin93) [reply lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java]: fixed

File: lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.expr.LambdaExpr;
+import javax.annotation.Nonnull;
+
+public final class NodeCaller implements Caller {
+
+    final Node callee;
+
+    public NodeCaller(@Nonnull final Node callee) {
```
[the remark this answers is not in the document]
Comment (ccostin93) [reply lib/src/main/java/com/iwillfailyou/javaparser/NodeCaller.java]: fixed.
Sorry, just got used to writing them on my day job.

**Comment by ccostin93** [comment]
@nikialeksey I cannot see why travis failed.
Running `./gradlew iwillfailyou check` locally passes for me.

**Comment by nikialeksey** [comment]
@ccostin93 All good!

Why build failed:
```
com.iwillfailyou.inspections.nomultiplereturn.multiplereturns.MultipleReturnViolationsTest > withTwoReturnsAndTwoInLambdaExpr FAILED

    java.lang.AssertionError at MultipleReturnViolationsTest.java:171

86 tests completed, 1 failed
```

Please, see why it failed, may be it is enviroment specific fail

**Comment by ccostin93** [comment]
> @ccostin93 All good!
> 
> Why build failed:
> 
> ```
> com.iwillfailyou.inspections.nomultiplereturn.multiplereturns.MultipleReturnViolationsTest > withTwoReturnsAndTwoInLambdaExpr FAILED
> 
>     java.lang.AssertionError at MultipleReturnViolationsTest.java:171
> 
> 86 tests completed, 1 failed
> ```
> 
> Please, see why it failed, may be it is enviroment specific fail

Took me a while to find the root cause since it never failed on my Mac, windows or linux machine :) On my travis it did fail randomly, ~1/3. The test was expecting 2 violations and then asserting that the description was correct for each one. The problem was that this was done with indexes on the list, but the order of the elements could not be guaranteed.

**Comment by ccostin93** [comment]
@nikialeksey did you drop the project?

**Comment by nikialeksey** [comment]
@ccostin93 I dont, but I had no motivation to continue, I have it now

**Pull request merged**